### DISCUSSION TITLE
How can I queue or asynchronously generate certificate and get a callback once certificate is generated?

### DISCUSSION THREAD

**Discussion by aks30** [Sunbird-RC/community#244] [General]
Is there an alternative approach to generate certificates apart from calling the generate API synchronously and waiting for the response so as to not worry about the Sunbird RC service being not available or delaying in giving a response in case of heavy load.

Cc - @aks30 @vijiurs @kiranharidas187 @varsha-sl @Prateek-slokam @aishwaryashikshalokam @priyanka-TL

**Comment by pramodkvarma** [comment]
Yes, that's how it was implemented in CoWin using Kafka. @dileepbapat can shed some light.

**Reply by coolbung** [reply]
Is the async feature instance level, or can it be enabled for select schemas only ?

**Reply by pramodkvarma** [reply]
My understanding is that both async and sync are available all the time. If you push some via async, those VC templates get executed, but if you call sync for other VC templates, it will generate those. Maybe @tejash-jl @dileepbapat can clarify.

**Reply by tejash-jl** [reply]
Right now it's at the instance level

**Reply by aks30** [reply]
So essentially at any given point we can have either sync or async, right?

**Reply by pramodkvarma** [reply]
@tejash-jl are you saying both sync API and async end points can be used together? Doesn't make sense. Can you explain why someone can't call the sync API even if async end point is enabled?

**Reply by tejash-jl** [reply]
We have not created a different API for async. When async is enabled the create entity API will choose the async mode to process the entities.

**Reply by pramodkvarma** [reply]
Ah! Got it. May be with the help of @tejash-jl, @aks30 can do a PR to allow both. Maybe simply through an additional API input param (optional param, when not passed, use the default config). Should be easy, I feel.

**Reply by aks30** [reply]
@tejash-jl I was going through the code here - https://github.com/Sunbird-RC/sunbird-rc-core/tree/main/services/certificate-api and couldn't find Kafka usage there, maybe if you can guide us what is the best way to enable both sync and async easily from the current setup.

**Reply by tejash-jl** [reply]
@aks30 https://github.com/Sunbird-RC/sunbird-rc-core/blob/main/java/registry/src/main/java/dev/sunbirdrc/registry/service/impl/RegistryAsyncServiceImpl.java#L34

**Reply by aks30** [reply]
@tejash-jl this one is for an entity where else we were looking for the certificate async and sync generation?

**Reply by aks30** [reply]
Hi, @tejash-jl is the async API presently available only for entities and not certificates?

**Reply by tejash-jl** [reply]
@aks30 
When an entity gets created, if there is a credential subject configured in the schema, we create a VC for that entity. The generated VC can be fetched as JSON/certificate (PDF/HTML/SVG). The complete flow is sync.
In async mode, when the create entity api is triggered, it pushes the entity to Kafka and the rest of the functionality gets completed by the consumer.

**Reply by aks30** [reply]
@tejash-jl What we were looking for is creating certificates against the existing entities in async mode. For example in a training certificate entity as and when different users complete their training for various things we want to generate certificates for them in async mode. Is that possible right now? Is it the same as an entity like at present i.e. at a time it can be working on sync or async mode? If so how do we make it work in both modes at the same time.

**Reply by sankarshanmukhopadhyay** [reply]
> @tejash-jl What we were looking for is creating certificates against the existing entities in async mode. For example in a training certificate entity as and when different users complete their training for various things we want to generate certificates for them in async mode. Is that possible right now? Is it the same as an entity like at present i.e. at a time it can be working on sync or async mode? If so how do we make it work in both modes at the same time.

I've been following this very interesting conversation (especially the use of a message queue etc) However, this model seems to be more of a choice between batch mode and just-in-time mode of generating and issuing certificates, is it not? A learner might complete training A today and receive a certificate and training B next week and receive another certificate then.

**Comment by tejash-jl** [comment]
@aks30,
yes, we have async functionality with webhook callback funcationality,
to enable async you need to set the below environment variables,
https://github.com/tejash-jl/ref-sunbird-certificate-demo/blob/async_mode/docker-compose.yml#L56-L60

also find the jupyter playbook for testing the apis, 
https://github.com/tejash-jl/ref-sunbird-certificate-demo/blob/async_mode/async-mode-api.ipynb

**Reply by aks30** [reply]
Thank you @tejash-jl I did check that but I couldn't find where do we specify topic name and any consumer group (in case multiple instances of RC are running to avoid duplicate message processing). Also if you can help with the Kafka message structure that we should use.

**Reply by tejash-jl** [reply]
@aks30,
the topic name is configurable, you can check it here,
https://github.com/Sunbird-RC/sunbird-rc-core/blob/main/java/registry/src/main/resources/application.yml#L62-L65

the group id is hardcoded and not configurable right now
https://github.com/Sunbird-RC/sunbird-rc-core/blob/main/java/registry/src/main/java/dev/sunbirdrc/registry/consumers/CreateEntityConsumer.java#L55

you can use the same POST api, the api pushes the input to kafka and creates it in async.

If you want to send the data directly to the topic, then below is the format
key:  `<unique UUID>`
message: 
```json
{
  "userId": "anonymous",
  "inputJson": {
    "TrainingCertificate": {
      "name": "Sunbird Learner",
      "contact": "79053896478",
      "trainingTitle": "Sunbird RC Certificate Module",
      "osOwner": [
        "anonymous"
      ]
    }
  },
  "skipSignature": false
}
```

i would suggest to use the API, instead of pushing directly to kafka topic

**Reply by aks30** [reply]
@tejash-jl Can you suggest why for the statement - "I would suggest to use the API, instead of pushing directly to Kafka topic", as it would mean we would have to handle failure and overload scenario in our side.

**Reply by tejash-jl** [reply]
just for clarification,
I meant to use the API with async_mode enabled so that the Sunbird-RC internally handles pushing it to the queue and also it calls the webhook for both success & failure events.

The other things that will be handled using the api:
- the input will be validated against the schema definition
- creation of user entities in keycloak (optional)

**Reply by pramodkvarma** [reply]
Definitely better to use API as it abstracts away queue/async tech from consumer. If we later allow adapters to Rabbit MQ etc, you will need to change code.

**Reply by aks30** [reply]
@pramodkvarma , @tejash-jl the benefit of using Kafka directly would be to handle the very scenario we mentioned in the first post i.e. in case RC service is down or under heavy load the message is there in the queue and RC can start processing once its available.

**Reply by pramodkvarma** [reply]
Yes, then you are making assumption about the actual format and Kafka as the pipe. If this is what you desire, you are better off managing Kafka on your side, push the message, and have a service that takes each message and call RC in async mode with retries etc. Using internal message format and bypassing RC async API is a bad idea from upgradeability etc.
